logic/LogicManager.java

```diff
@@ -18,43 +18,133 @@
+import grakn.core.traversal.TraversalEngine;
+import graql.lang.pattern.Conjunctable;
 import graql.lang.pattern.Conjunction;
+import graql.lang.pattern.Negation;
 import graql.lang.pattern.Pattern;
+import graql.lang.pattern.constraint.TypeConstraint;
+import graql.lang.pattern.variable.BoundVariable;
 import graql.lang.pattern.variable.ThingVariable;
 
+import java.util.Optional;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static grakn.core.common.exception.ErrorMessage.RuleWrite.TYPES_NOT_FOUND;
+
 public class LogicManager {
 
     private final ConceptManager conceptMgr;
     private final GraphManager graphMgr;
     private final TypeHinter typeHinter;
+    private LogicCache logicCache;
 
-    public LogicManager(GraphManager graphMgr, ConceptManager conceptMgr, TypeHinter typeHinter) {
+    public LogicManager(GraphManager graphMgr, ConceptManager conceptMgr, TraversalEngine traversalEng, LogicCache logicCache) {
         this.graphMgr = graphMgr;
         this.conceptMgr = conceptMgr;
-        this.typeHinter = typeHinter;
+        this.logicCache = logicCache;
+        this.typeHinter = new TypeHinter(conceptMgr, traversalEng, logicCache.hinter());
     }
 
     public Rule putRule(String label, Conjunction<? extends Pattern> when, ThingVariable<?> then) {
-        final RuleStructure structure = graphMgr.schema().getRule(label);
-        if (structure != null) structure.delete();
-        Rule rule = Rule.of(conceptMgr, graphMgr, typeHinter, label, when, then);
-
-        // TODO detect negated cycles in the rule graph after inserting this rule and updating a rule graph
-
-        return rule;
+        RuleStructure structure = graphMgr.schema().getRule(label);
+        if (structure != null) {
+            // overwriting a rule means we purge it and re-create the rule
+            structure.delete();
+            logicCache.rule().invalidate(label);
+        }
+        return logicCache.rule().get(label, l -> Rule.of(graphMgr, conceptMgr, this, label, when, then));
     }
 
     public Rule getRule(String label) {
-        final RuleStructure structure = graphMgr.schema().getRule(label);
-        if (structure != null) return Rule.of(conceptMgr, structure, typeHinter);
+        Rule rule = logicCache.rule().getIfPresent(label);
+        if (rule != null) return rule;
+        RuleStructure structure = graphMgr.schema().getRule(label);
+        if (structure != null) {
+            return logicCache.rule().get(structure.label(), l -> Rule.of(this, structure));
+        }
         return null;
     }
 
-    public TypeHinter typeHinter() {
+    public ResourceIterator<Rule> rules() {
+        return graphMgr.schema().rules().map(struct -> {
+            Rule rule = logicCache.rule().getIfPresent(struct.label());
+            if (rule == null) rule = logicCache.rule().get(struct.label(), l -> Rule.of(this, struct));
```

Comment:
We check nulls everywhere in the storage layer rather than use `Optional<>`s?

Reply:
yep

Reply:
because rocks can return null values for a key

Reply:
how can RocksDB return null values for a key, actually, @flyingsilverfin ?